packages/lib/models/Note.js

```diff
@@ -401,7 +410,13 @@ class Note extends BaseItem {
 			options.conditions.push('is_todo = 1');
 		}
 
-		return this.search(options);
+		const results = await this.search(options);
+
+		if (options.order.length > 0 && options.order[0].by == 'title' && Setting.value('titleNaturalSort')) {
+			results.sort((a, b) => this.naturalSortCompare(a.title, b.title, options.order[0].dir));
+		}
```

Comment:
This part of Joplin is used by all apps, plugins and pretty much all services, so I'm a bit concerned about the performance penalty. Up to now, sorting was done at the database level, while now it would be done in code

Is it not possible to implement this at the database level again (in BaseModel.applySqlOptions)? If not possible, please could you provide some info about the performance of using this function vs not using it? Let's say when processing ten thousand notes - how long does it take to sort them?

Reply:
Also, a minor issue, but there shouldn't be duplicate code. So the function should be refactored or sorting part moved to a separate function and called from here.

Reply:
- I understand. I initially wanted to do natural sorting at database level as well, the usual way handles numbers by their length, e.g., `SELECT item_no FROM items ORDER BY LENGTH(item_no) , item_no;`, but that cannot correctly sort titles like `A10` and `A9:good`, because `A9:good` is longer. Also we sort notes (1) when querying database and (2) in sortNotes() function, where we need to perform natural sort in javascript anyway.

Now comes to performance. I did a test by inserting 10K notes with random titles and sort them. The way I used initially is `localeCompare()`, its performance is poor. It took ~400ms to complete previews() function whereas it only took ~60ms to complete without that.

The good news is that [defining a `Intl.Collator` before sorting vastly improved the performance](https://stackoverflow.com/questions/14677060/400x-sorting-speedup-by-switching-a-localecompareb-to-ab-1ab10). It took only ~70ms to complete previews() function, compared with ~60ms to complete without that.

- Regarding the duplicate code related to sorting, I put them in a separate function now.

Reply:
That makes sense, and in that case let's do sorting in code then. Thanks for looking into this!